### PR TITLE
`to` function for converting the backend TensorMap values torch <-> numpy

### LINKED ISSUE #70: Create a utility function to convert ``TensorMap`` block values to type ``torch.tensor``
In order to be compatible as input to PyTorch models, the ``TensorBlock.values`` tensors in each block of a ``TensorMap`` need to be of type ``torch.tensor``, instead of a ndarray, i.e. ``numpy.ndarray`` or ``_RustNDArray``.

**Desired Behaviour**

There should be a utility function accessible through the Python API, perhaps in a new file such as _equistore/python/src/equistore/operations/tensormap_to_torch.py_, that is used in the following way:
```py
new_tensormap = tensormap_to_torch(tensormap)
```

**Implications**

This necessarily has to be implemented as part of the Python API as there is no Rust-side solution to creating ``torch.tensor`` objects. This means that the above function ``tensormap_to_torch()`` will have to return a new ``TensorMap`` object, thus creating a copy of the object in memory. However, doing it this way round means that all the infrastructure that builds descriptors need not be written to be compatible with PyTorch; only at the step before building a model do we need to convert our blocks to ``torch.tensor`` objects.

### PR DESCRIPTION
**Pull request by jwa7**
Closes #70 and #149.

Still to do:

- [x] Finish off functionality
- [x] Write unit tests

To discuss:

- Let the arg `dtype` be passed as a `torch.dtype` object, or as a`str`? The former is more flexible.
- Should the `detach()` function just be a wrapper for `to`? Do we need a `detach` function as well as `to`?
- Add device support (and maybe dtype support) for the private functions `_to_numpy` and `_to_numpy_block`?

### CODE REVIEW

File: python/src/equistore/operations/to.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+import numpy as np
+import torch
+
+from ..block import TensorBlock
+from ..labels import Labels
+from ..tensor import TensorMap
+
+
+def to(
```
Comment (Luthaf) [inline python/src/equistore/operations/to.py:9]: I would rather have two separate functions for TensorMap and TensorBlock, like the rest of the operations API

File: python/src/equistore/operations/to.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            f"``block`` should be an equistore `TensorBlock`, got {type(block)}"
+        )
+    if not isinstance(backend, str):
+        raise TypeError("`backend` should be passed as a `str`")
+    if backend == "torch":
+        # Check args
+        if not isinstance(requires_grad, bool):
+            raise TypeError("`requires_grad` should be passed as a `bool`")
+        if not isinstance(dtype, torch.dtype):
+            raise TypeError("`dtype` should be passed as a `torch.dtype`")
```
Comment (Luthaf) [inline python/src/equistore/operations/to.py:93]: Not really, torch also accepts strings here. Overall, I think it is fine not to type check these, and pass them through to the backend which will handle them for us.

File: python/src/equistore/operations/to.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        if not isinstance(device, torch.device):
+            raise TypeError("`device` should be passed as a `torch.device`")
+        return _to_torch_block(block, requires_grad, dtype, device)
+    elif backend == "numpy":  # numpy
+        # TensorBlock
+        return _to_numpy_block(block)
```
Comment (Luthaf) [inline python/src/equistore/operations/to.py:99]: this should also take a dtype

File: python/src/equistore/operations/to.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    block: TensorBlock,
+    requires_grad: bool,
+    dtype: torch.dtype,
+    device: torch.device,
+) -> TensorBlock:
+    """
+    Creates a new :py:class:`TensorBlock` where block values are PyTorch
+    :py:class:`torch.tensor` objects. Assumes the block values are already as a
+    type that is convertible to a :py:class:`torch.tensor`, such as a numpy
+    array or RustNDArray. The resulting torch tensor dtypes are enforced as
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply python/src/equistore/operations/to.py:114]: RustNDArray does not exists anymore

File: python/src/equistore/operations/to.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    Takes a TensorBlock object whose values are torch.tensor objects and
+    converts them to numpy arrays of dtype np.float64. Returns a new TensorBlock
+    object.
+    """
+    if isinstance(block.values, np.ndarray):
+        return block.copy()
+
+    # Create new block, with the values tensor converted to a torch tensor.
+    new_block = TensorBlock(
+        values=np.array(block.values.detach().numpy(), dtype=np.float64),
```
Comment (Luthaf) [inline python/src/equistore/operations/to.py:163]: I don't love that we assume torch here, but I think we have to since we need to call detach =/

Do you know if we could use dlpack functionalities here instead?

File: python/src/equistore/operations/to.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        components=block.components,
+        properties=block.properties,
+    )
+
+    # Add gradients to each block, again where the values tensor are torch
+    # tensors.
+    for parameter, gradient in block.gradients():
+        new_block.add_gradient(
+            parameter,
+            np.array(gradient.data.detach().numpy(), dtype=np.float64),
```
Comment (jwa7) [inline python/src/equistore/operations/to.py:174]: This will need updating to reflect the new Gradients as TensorBlocks change

File: python/src/equistore/operations/to.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+
+def _to_numpy_block(block: TensorBlock) -> TensorBlock:
+    """
+    Takes a TensorBlock object whose values are torch.tensor objects and
+    converts them to numpy arrays of dtype np.float64. Returns a new TensorBlock
+    object.
+    """
+    if isinstance(block.values, np.ndarray):
+        return block.copy()
```
Comment (jwa7) [inline python/src/equistore/operations/to.py:159]: What memory considerations should we make here @Luthaf @frostedoyster ? Do we want to provide the option to release the previous tensor from memory? If so, should the default behaviour be safe (return a copy) or unsafe?

See discussion in https://github.com/lab-cosmo/equistore/pull/248

File: python/src/equistore/operations/to.py (hunk 5 above)
Comment (frostedoyster) [reply python/src/equistore/operations/to.py]: I removed the detach, not sure if that still works. If necessary, we can only detach if values is an instance of torch.Tensor

Comment (Luthaf) [reply python/src/equistore/operations/to.py]: > If necessary, we can only detach if values is an instance of torch.Tensor

Yes, we'll have to detach torch tensor, otherwise we'll get an error in the call to `.numpy()`

File: python/equistore-operations/tests/to.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    block = equistore.block_from_array(values)
+    grad_block = equistore.block_from_array(grad_values)
+    grad_grad_block = equistore.block_from_array(grad_grad_values)
+    grad_block.add_gradient("grad_grad", grad_grad_block)
+    block.add_gradient("grad", grad_block)
+    return block
+
+
+@pytest.fixture
+def tmap():
```
Comment (Luthaf) [inline python/equistore-operations/tests/to.py:25]:
```suggestion
def tensor():
```

Please optimize code for readability, not ease of writing. Much more time is spent reading code than writing it: https://www.goodreads.com/quotes/835238-indeed-the-ratio-of-time-spent-reading-versus-writing-is

Comment (frostedoyster) [reply python/equistore-operations/tests/to.py:25]: Yes but that doesn't return a tensor

Comment (Luthaf) [reply python/equistore-operations/tests/to.py:25]:
```
    tmap = TensorMap(keys=Labels.arange("dummy", 2), blocks=[block1, block2])
    return tmap
```

Comment (Luthaf) [reply python/equistore-operations/tests/to.py:25]: We use `tensor` for all fixtures returning a TensorMap: https://github.com/lab-cosmo/equistore/blob/7e8b41926ca61039f9113510daa8191270aa23c0/python/equistore-operations/tests/add.py#L15

**Comment by jwa7** [comment]
I've made some changes - using dispatch, merged master, and some updates to docs, typing and style. I've broken the CI so will fix next.

Just a couple of questions about the default value of `requires_grad`. If I have a torch-based TensorMap and I use `to` with the backend set to "torch". At the moment, in this case, the `requires_grad` arg is essentially ignored.

1) Should we allow the user, in the above example, to change `requires_grad`?
2) Should the default value of `requires_grad` be `None` instead?

**Review by frostedoyster — COMMENTED** [review]
Is that return necessary? Regarding the requires_grad for the torch-to-torch conversion, I don't know... actually the only case where it is useful is probably when going from numpy to torch. We could also remove it altogether and then people would have to modify requires_grad after calling `equistore.to`

File: python/equistore-operations/equistore/operations/_dispatch.py [hunk 9]
```diff
@@ -400,3 +400,46 @@ def rand_like(array, shape=None, requires_grad=False):
...
+        if backend == "numpy":
+            return array.detach().cpu().numpy()
+
+        elif backend == "torch":
+            # We need this to keep gradients of the tensor
+            return array.to(dtype=dtype, device=device)
+
+        else:
+            raise ValueError(f"Unknown backend: {backend}")
+        return np.array(array, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/equistore-operations/equistore/operations/_dispatch.py]: Re: requires_grad - yes that makes sense. Thoughts @Luthaf ?

[the remark this answers is not in the document]
Comment (Luthaf) [reply python/equistore-operations/equistore/operations/_dispatch.py]: I would agree that we could leave it off, and add a separate operation to change requires_grad (which should be doable without creating new TensorMap/TensorBlock)

File: python/equistore-operations/equistore/operations/_dispatch.py [hunk 10]
```diff
@@ -402,7 +402,7 @@ def rand_like(array, shape=None, requires_grad=False):
         raise TypeError(UNKNOWN_ARRAY_TYPE)
 
 
-def to(array, backend: str = None, dtype=None, device=None, requires_grad=False):
+def to(array, backend: str = None, dtype=None, device=None, requires_grad=None):
```
Comment (Luthaf) [inline python/equistore-operations/equistore/operations/_dispatch.py:405]: I think it would be nicer long term to have separate `to_torch` and `to_numpy` functions, taking whichever parameter they need to. This makes it easier to add other backends later.

**Review by Luthaf — COMMENTED** [review]
Only one small comment and this should be good to go
